torch/vmap_internals.py

```diff
@@ -0,0 +1,89 @@
+import torch
+import functools
+from torch import Tensor
+import warnings
+
+REQUIRE_SAME_MAP_SIZE = (
+    'vmap: Expected all tensors to have the same size in the mapped dimension, '
+    'got sizes {sizes} for the mapped dimension'
+)
+
+ELEMENT_MUST_BE_TENSOR = (
+    'vmap({fn}, ...): `{fn}` must return a Tensor or a flat sequence of tensors, got '
+    'type {out} for return {idx}.'
+)
+
+MUST_BE_TENSOR_LIST = (
+    'vmap({fn}, ...): `{fn}` must return a Tensor or a flat sequence of tensors, got '
+    'type {out} as the return.'
+)
+
+NO_INPUTS = (
+    'vmap({fn})(<inputs>): got no inputs. Maybe you forgot '
+    'to add inputs, or you are trying to vmap over a '
+    'function with no inputs. The latter is unsupported.'
+)
+
+# Checks that all args have the same batch dim size.
+def _validate_and_get_batch_size(args):
+    batch_sizes = [arg.size(0) for arg in args]
+    if batch_sizes and any([size != batch_sizes[0] for size in batch_sizes]):
+        raise ValueError(REQUIRE_SAME_MAP_SIZE.format(sizes=batch_sizes))
+    return batch_sizes[0]
+
+
+def _validate_inputs_and_get_batch_size(args, fn_name):
+    if len(args) == 0:
+        raise ValueError(NO_INPUTS.format(fn=fn_name))
+    return _validate_and_get_batch_size(args)
+
+# Undos the batching (and any batch dimensions) associated with the `vmap_level`.
+def _unwrap_batched(batched_outputs, vmap_level, batch_size):
+    if isinstance(batched_outputs, Tensor):
+        return torch._remove_batch_dim(batched_outputs, vmap_level, batch_size, 0)  # type: ignore
```

Comment:
OOC, why do you need to `type: ignore` here? We should have accurate type annotations for these functions!

Reply:
I think the answer is mypy doesn't like that we're calling methods on torch while we `import torch` (this file gets imported as a part of the `import torch` process) but I didn't dig very far into the error. I'll dig a little and then put a comment to that effect here if my hypothesis is supported

Reply:
Update: I spent some time looking into this. mypy is ignoring all functions on torch that begin with an underscore. I don't know why this is (debugging the type checker is interestingly difficult) but I filed an issue at #40397. I'll just link to that issue from an in-line comment here.

Reply:
One way that might work to workaorund this problem is to say `torch._C._VariableFunctions._remove_batch_dim` instead. (Note that this workaround doesn't work if you're TorchScripting)